### PR TITLE
v0.2.0

### PR DESCRIPTION
**Pull request by julman99**
## Summary by CodeRabbit

* **New Features**
  * Percentage-based memory sizing, improved automatic chunk sizing, and clearer CLI flow.

* **Improvements**
  * Streamlined Docker images and multi-architecture build/publish workflows; deterministic build versioning and simplified runtime image.
  * Cross-platform memory detection on Linux, macOS, and Windows for accurate sizing.

* **Documentation**
  * Complete README rewrite with expanded usage, examples, options, and exit-code docs.

* **Testing**
  * Expanded cross-platform and compiler test suites with runnable test scripts.

* **Chores**
  * Docker build context tightened to whitelist only essential files.

[Review Change Stack](https://app.coderabbit.ai/change-stack/julman99/eatmemory/pull/17)

### CODE REVIEW

File: .cursor/rules/kluster-code-verify.mdc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+alwaysApply: true
+---
+
+Always run kluster_bug_check_tool each time when new code is generated or existing code updated. Follow kluster_bug_check_tool fields description strictly.
+You must always execute and strictly follow todo list from "agent_todo_list" returned by any kluster tool, if it exists. Do not stop until there is something left in "agent_todo_list". Do all the fixes from "agent_todo_list" before running "kluster_bug_check_tool" again.
```
[the remark this answers is not in the document]
Comment (julman99) [reply .cursor/rules/kluster-code-verify.mdc]: The `.cursor/rules/kluster-code-verify.mdc` file was removed from the branch in a history rewrite; this comment no longer applies.

File: .cursor/rules/kluster-code-verify.mdc [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+Always run kluster_frameworks_check_tool before adding new package or running package manager such as npm, gradle, pip etc. Also, run when it directly generated or updated file with list of packages, such as package.json
```
[the remark this answers is not in the document]
Comment (julman99) [reply .cursor/rules/kluster-code-verify.mdc]: Same as the adjacent comment — file removed in the history rewrite, no longer applies.

File: src/eatmemory.c [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+// compiler cannot constant-fold the write loop or the verification loop.
+static inline uint8_t eatmemory_pattern(size_t chunk_index, size_t byte_offset, size_t total) {
+    return (uint8_t)((chunk_index + byte_offset) ^ total);
+}
+
+struct allocation eat(size_t total, size_t chunk_size, eatmemory_error* error) {
+    struct allocation result = { NULL, 0 };
+    *error = EM_ERROR_NONE;
+
+    size_t iterations = total/chunk_size;
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.c:175]: Fixed in `f4b5e99` — `eat()` now rejects `chunk_size == 0` up front with `EM_ERROR_CHUNK_SIZE_ARG_INVALID`, and main.c maps the error to exit 11. Regression test added in `test/simple.test.sh`.

File: src/eatmemory.c [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+#define MIN(a,b) ((a) < (b) ? (a) : (b))
+#define MAX_VALUE_STR_SIZE 255
+
+#define TO_KB 1024UL
+#define TO_MB (1024UL * TO_KB)
+#define TO_GB (1024UL * TO_MB)
+
+size_t string_to_bytes(char * str, eatmemory_error* error) {
+    const size_t len = strlen(str);
+    char unit = str[len - 1];
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.c]: Fixed in `f4b5e99` — `string_to_bytes` now rejects `len == 0` and `len >= MAX_VALUE_STR_SIZE` up front before any indexing into `str`. Regression test added in `test/simple.test.sh`.

File: src/eatmemory.c [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    char value_numeric[MAX_VALUE_STR_SIZE];
+
+    strncpy(value_numeric, str, MAX_VALUE_STR_SIZE);
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.c]: Fixed in `f4b5e99` — replaced `strncpy` with `memcpy(value_numeric, str, len + 1)` which explicitly copies the NUL terminator, plus the length check ensures the buffer fits.

File: test/simple.test.sh [hunk 6]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+# simple.test.sh - Simple test that compiles and runs eatmemory with edge cases around MIN_VERIFICATION_THRESHOLD_BYTES
+# Tests memory allocations below, at, and above the 1MB verification threshold
+
+set -e  # Exit on any error
+
+# Constants for edge case testing around MIN_VERIFICATION_THRESHOLD_BYTES
+readonly THRESHOLD_BYTES=1048576                           # 1MB in bytes
+readonly THRESHOLD_KB=$((THRESHOLD_BYTES / 1024))          # 1024KB
```
[the remark this answers is not in the document]
Comment (julman99) [reply test/simple.test.sh]: Fixed in `375800d` — renamed the `THRESHOLD_*` constants to neutral names and rewrote the section header and `expected_behavior` strings to describe what's actually being exercised (the 1 MB boundary in `get_auto_chunk_size` and `bytes_to_string`).

File: src/eatmemory.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
+#include <stdio.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "errors.h"
+
+#ifndef eatmemory_h
+#define eatmemory_h
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.h]: Fixed in `969a497` — the four `#include` lines were moved inside the `#ifndef eatmemory_h` guard.

File: src/eatmemory.c [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    // leave value_numeric un-terminated when strlen(str) >= MAX_VALUE_STR_SIZE,
+    // which we have ruled out above but the explicit copy is clearer.
+    memcpy(value_numeric, str, len + 1);
+    if(!isdigit(unit)) {
+        value_numeric[len - 1] = '\0';
+    }
+
+    //parse bytes into numeric variable
+    size_t bytes;
+    if(sscanf(value_numeric, "%zu", &bytes) == 0) {
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.c]: Fixed in `6f0c138` — `sscanf` check changed from `== 0` to `!= 1`, catching both EOF and 0. Regression tests for `eatmemory M` and `eatmemory %` added to `test/simple.test.sh`.

File: test/simple.test.sh [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    fi
+    echo "✓ Test $test_num completed (exit $actual_exit as expected)"
+}
+
+echo ""
+echo "Testing edge cases around MIN_VERIFICATION_THRESHOLD_BYTES (${THRESHOLD_KB}KB = ${THRESHOLD_BYTES} bytes):"
+echo "=================================================================================================="
+
+# Test cases around the verification threshold
+run_test "1" "Below threshold (${THRESHOLD_MINUS_1_KB}KB)" "${THRESHOLD_MINUS_1_KB}K -t 0" "Memory verification should be disabled"
```
[the remark this answers is not in the document]
Comment (julman99) [reply test/simple.test.sh]: Fixed in `375800d` — same fix as the duplicate Greptile thread; the `MIN_VERIFICATION_THRESHOLD_BYTES` references were removed and the test narrative rewritten.

File: src/main.c [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+#include <ctype.h>
+#include <stdbool.h>
+#include "eatmemory.h"
+#include "args.h"
+#include <unistd.h>
+#include "errors.h"
+
+ArgParser* configure_cmd() {
+    ArgParser* parser = ap_new_parser();
+    ap_add_flag(parser, "help h ?");
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/main.c:25]: Skipped. Two reasons:

1. The library's `ap_set_helptext` only auto-handles `--help`/`-h`; the project also exposes `?` as a short form, which would be lost under the library's auto-mode.
2. `print_help()` does conditional formatting (the `%` line is gated on `SYSMEM_MODE_UNKNOWN`, it prints `VERSION`, and it formats `bytes_to_string(SIZE_MAX, ...)` dynamically). Pre-computing a static string to hand to `ap_set_helptext` is possible but adds machinery for no user-visible win.

Leaving as-is.

File: src/main.c [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    printf("Options:\n");
+    printf("-t <seconds>     Exit after specified number of seconds.\n");
+    printf("-s <chunk_size>  Specify a custom chunk size in the same format\n");
+    printf("                 as the memory to be eaten.\n");
+    printf("                 Default: %s\n", STR_CHUNK_AUTO);
+    printf("                 Max:     %s\n", bytes_to_string(SIZE_MAX, BYTES_TMP_STR()));
+    printf("\n");
+}
+
+void print_and_exit(char * error, eatmemory_error exit_code) {
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/main.c:51]: Skipped — `exit_with_error` is `static` in `args.c` and not exposed, so the rule as stated can't be followed. The adjacent real improvement (errors should go to stderr, not stdout) was applied separately in `70f3aca`.

File: src/eatmemory.c [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        size_t denominator = 1;
+        if (unit == 'K') {
+            numerator = TO_KB;
+        } else if (unit == 'M') {
+            numerator = TO_MB;
+        } else if (unit == 'G') {
+            numerator = TO_GB;
+        } else if (unit == '%') {
+            struct system_memory_stats memory_stats;
+            get_system_memory_stats(&memory_stats);
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.c:86]: Valid design point, deferred. Logged as #18 for a future refactor. The current coupling has no user-visible effect and would touch the parser's signature plus ripple into main.c — out of scope for this maintenance branch.

File: src/eatmemory.c [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        // Overflow gate for `bytes * numerator`. The numerator > 0 guard
+        // covers '%' where the numerator is runtime-derived and could
+        // theoretically be zero (system reports no free memory).
+        if (numerator > 0 && bytes > SIZE_MAX / numerator) {
+            *error = EM_ERROR_PARSE_OVERFLOW;
+            return 0;
+        }
+        // Multiply first, then divide -- preserves precision for small
+        // values of `bytes` in the '%' case.
+        bytes = bytes * numerator / denominator;
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.c:108]: No longer applicable after the current parser implementation. Unit math is now done in `uint64_t`, with overflow checked against `UINT64_MAX` before multiplication and final narrowing checked against `SIZE_MAX` after division. The specific 32-bit case described here no longer trips a `size_t` intermediate overflow.

File: src/args.c [hunk 14]
```diff
@@ -1109,18 +1109,24 @@ static void ap_handle_equals_opt(ArgParser* parser, const char* prefix, const ch
     bool found = map_get(parser->option_map, name, (void**)&option);
 
     if (!found) {
+        char error_msg[256];
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/args.c:1113]: Fixed in `e571330`: `str()` now calls `va_end(args)` before returning on the `vsnprintf` formatting-error path.

[the remark this answers is not in the document]
Comment (julman99) [reply src/args.c:1113]: Fixed in `e571330`: the command registration rollback now decrements `parent_parser->command_vec->count` instead of moving the `Vec*` pointer.

File: src/eatmemory.c [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+    memcpy(value_numeric, str, len + 1);
+    if(!isdigit((unsigned char)unit)) {
+        value_numeric[len - 1] = '\0';
+    }
+
+    // Parse and apply unit math in uint64_t, then narrow to size_t at the
+    // return. This lets 32-bit platforms compute '50% of 2 GB' (a 100 GB
+    // intermediate, 1 GB final) without spuriously rejecting at the
+    // multiplication stage just because the intermediate exceeds size_t.
+    if (value_numeric[0] == '-') {
```
[the remark this answers is not in the document]
Comment (julman99) [reply src/eatmemory.c]: Fixed in `84a778c`: `string_to_bytes()` now trims leading input before checking for `-`, parses with `strtoumax`/`endptr`, and validates the optional unit after the parsed number. Added a regression for `" -1M"`.
